File: IntToWord.java
/*
  time : O(n)
  Space: O(1)
  Run on LeetCode: yes
*/
class Solution {

    String ones[] = {"", "One ", "Two ", "Three ", "Four ", "Five ", "Six ", "Seven ", "Eight ", "Nine ", "Ten "};
    String teen[] = {"Ten ", "Eleven ", "Twelve ", "Thirteen ", "Fourteen ", "Fifteen ", "Sixteen ", "Seventeen ", "Eighteen ", "Nineteen "};
    String normal[] = {"","","Twenty ", "Thirty ", "Forty ", "Fifty ", "Sixty ", "Seventy ", "Eighty ", "Ninety "};
    String cent[] = {"", "Thousand ", "Million ", "Billion "};

    private StringBuilder createHunder(int num){
        StringBuilder last = new StringBuilder();

        int cnt = 0;
        int temp = num;
        while(num > 0){
            int n = num %10;
            num /= 10;
            StringBuilder str = new StringBuilder();
            if(cnt == 1){
                if(n == 1){
                    int tn = temp%10;
                    last = new StringBuilder();
                    str .append(teen[tn]);
                }
                else{
                    str.append(normal[n]);
                }
            }
            else if(cnt == 2){
                str.append(ones[n]);
                if(n > 0){
                    str.append("Hundred ");
                }
            }
            else{
                str.append(ones[n]);
            }
            str.append(last);
            last = str;
            cnt++;
        }
        return last;
    }

    public String numberToWords(int num) {
        if(num == 0)
            return "Zero";
        int cen = 1000;
        StringBuilder last = new StringBuilder();
        int indx = 0;
        while(num > 0){
            //get last hunderd int.
            int hnd = num % cen;
            num = num/cen;

            //convert lasthunverd to a hundred.
            StringBuilder str = new StringBuilder();
            str = createHunder(hnd);

            //add centruy level.
            if(str.length() > 0)
                str.append(cent[indx]);
            str.append(last);

            //set this str as last.
            last = str;
            indx++;
        }
        last.deleteCharAt(last.length()-1);
        return last.toString();
    }
}

File: IntToWord.py
class Solution:
    ones = ["", "One ", "Two ", "Three ", "Four ", "Five ", "Six ", "Seven ", "Eight ", "Nine ", "Ten "]
    
    ninteen = ["Twenty ", "Eleven ", "Twelve ", "Thirteen ", "Fourteen ", "Fifteen ", "Sixteen ", "Seventeen ", "Eighteen ", "Nineteen ", "Twenty "]
    
    tens = [" ", "Ten ", "Twenty ", "Thirty ", "Forty ", "Fifty ", "Sixty ", "Seventy ", "Eighty ", "Ninety ", "Hundred "]
    
    thousands = [" ", "Thousand ", "Million ", "Billion "]
    
    def helper(self, x):
        
        h = x // 100
        r = x % 100
        result = ""
        
        if h != 0:
            result = result + self.ones[h] + "Hundred "
        
        #print(r)
        if r <= 20:
            if r >= 0 and r <= 10:
                result = result + self.ones[r]
            
            elif r >= 11 and r <= 20:
                result = result + self.ninteen[r % 10]
        else:
            t = r // 10
            new_r = r % 10
            
            #print(t, new_r)
            result = result + self.tens[t]
            result = result + self.ones[new_r]
        #print(result)
        return result
        
    def numberToWords(self, num: int) -> str:
        if num == 0:
            return "Zero"
        
        temp_list = []
        result = ""
        while num > 0:
            temp_list.append(num % 1000)
            num = num // 1000
        
        print(temp_list)
        while len(temp_list) > 0:
            i = temp_list.pop()
            a = self.helper(i)
            #print(a)
            if a!= "":
                result = result + a + self.thousands[len(temp_list)]
            #print(result)
        #print(result)
        #self.helper(570)
        return result.strip()
